adapters/mobfoxpb/mobfoxpb.go

```diff
@@ -0,0 +1,137 @@
+	return &MobfoxpbAdapter{
+		URI: endpoint,
+	}
+}
+
+// MakeRequests create bid request for mobfoxpb demand
+func (a *MobfoxpbAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var err error
+	var tagID string
+
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb.Imp{imp}
+
+		tagID, err = jsonparser.GetString(reqCopy.Imp[0].Ext, "bidder", "TagID")
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		reqCopy.Imp[0].TagID = tagID
+
+		adapterReq, errors := a.makeRequest(&reqCopy)
+		if adapterReq != nil {
+			adapterRequests = append(adapterRequests, adapterReq)
+		}
+		errs = append(errs, errors...)
+	}
+	return adapterRequests, errs
+}
+
+func (a *MobfoxpbAdapter) makeRequest(request *openrtb.BidRequest) (*adapters.RequestData, []error) {
+
+	var errs []error
+
+	reqJSON, err := json.Marshal(request)
+
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.URI,
+		Body:    reqJSON,
+		Headers: headers,
+	}, errs
+}
+
+// MakeBids makes the bids
+func (a *MobfoxpbAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusNotFound {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			bidType, err := getMediaTypeForImp(sb.Bid[i].ImpID, internalRequest.Imp)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				b := &adapters.TypedBid{
+					Bid:     &sb.Bid[i],
+					BidType: bidType,
+				}
+				bidResponse.Bids = append(bidResponse.Bids, b)
+			}
+		}
+	}
+	return bidResponse, errs
+}
+
+func getMediaTypeForImp(impID string, imps []openrtb.Imp) (openrtb_ext.BidType, error) {
+	mediaType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impID {
+			if imp.Banner == nil && imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
```

Comment:
Does this mean that for multi-format requests i.e. requests that would have an imp with both banner and video defined, your server is always guaranteed to return a banner response?

Reply:
Does that mean your system discards requests that have multiple formats or that you always just consider the `banner` part in that request and return bids just for that?

Reply:
first bid.imp (bid.imp[0]) will be taken into account

Reply:
Do you mean `request.Imp[0]`? Even that single `imp` can have both `banner` and `video` set for multi-format requests. In such a case if you always only consider the banner part of the imp and bid only on that then this code is fine but if you consider both banner and video and have some logic to determine which one to bid on then you could be sending a video bid back but marking it as `banner` here. I hope that makes sense.

Reply:
Something should be set as default, media-type can't be empty, and even in case some one will send us multiformat impression - key, passed in extension will lead to predefined endpoint with predefined media type, so response will have expected type

Reply:
What we're asking for here is to either read the type of the format from your bidding server's response or perfectly duplicate the logic from your server here. If your server gets an imp for both a banner and video, which will it choose?

Reply:
We are not expecting to get multiformat impression. Our backend logic looks something like that:
```
if(bid.banner) return 'banner'
if(bid.video) return 'vide'
```

Reply:
so if imp has banner - that would be banner traffic, even if video there inside bid

Reply:
actualy the code in adapter do the same

Reply:
Given that we are not null-checking `imp.Banner` nor `imp.Video` anywhere else in this adapter code, I believe the truth table below describes the scenarios we may find. Right?
```
  | imp.Banner  | imp.Video   | other (imp.Audio OR imp.Native)  ||  RESULT
--|-------------|-------------|----------------------------------||---------------------------
 1| nil         | nil         | nil                              || openrtb_ext.BidTypeBanner
 2| nil         | nil         | not-nil                          || openrtb_ext.BidTypeBanner
 3| nil         | not-nil     | nil                              || openrtb_ext.BidTypeVideo
 4| nil         | not-nil     | not-nil                          || openrtb_ext.BidTypeVideo
 5| not-nil     | nil         | nil                              || openrtb_ext.BidTypeBanner
 6| not-nil     | nil         | not-nil                          || openrtb_ext.BidTypeBanner
 7| not-nil     | not-nil     | nil                              || openrtb_ext.BidTypeBanner
 8| not-nil     | not-nil     | not-nil                          || openrtb_ext.BidTypeBanner
```

If so, are we ok with this logic? Or should we modify and set to `BidTypeVideo` in scenarios 7, and 8? Should we error out in the somewhat unlikely case scenarios 1 or 2 happen?

Reply:
I've made some changes to be closer to our backend logic, so for now we check mediatypes in such way (on backend and in pbs adapter) first we check banner, if no - we check video, if no - we check native, at least one media type will be found, if we recieve unsuported media type - bid would be rejected.  About #1 - pbs will reject any request without at least one media type.